src/Adapter/Attribute/Repository/AttributeRepository.php

```diff
@@ -82,4 +84,128 @@ public function assertAllAttributesExist(array $attributeIds): void
             throw new AttributeNotFoundException('Some of provided attributes does not exist');
         }
     }
+
+    /**
+     * @param ProductId $productId
+     * @param LanguageId $langId
+     *
+     * @return array[]
+     */
+    public function getAttributesInfoByProductId(ProductId $productId, LanguageId $langId): array
+    {
+        $associatedProductAttributes = $this->getProductAttributes($productId->getValue());
+        $attributeIds = array_unique(array_map(function (array $attributeByCombination): int {
```

Comment:
It will maybe be faster to use a new methods `$this->getProductAttributesIds($productId->getValue());` for performance, I'm afraid of side effect if you have 1000 combinations, array_map is rarely fast

Reply:
I see that you use `getProductAttributes` only in this class, why not doing the uniq check in SQL, it will be faster